File: cmd/secureyeoman-edge/llm.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"
)

// isPrivateIP returns true when rawURL's host resolves to a private, loopback,
// or link-local address. Unparseable URLs are treated as private (blocked).
func isPrivateIP(rawURL string) bool {
	u, err := url.Parse(rawURL)
	if err != nil {
		return true // block unparseable URLs
	}
	host := u.Hostname()
	// Block obvious private hosts
	if host == "localhost" || host == "metadata.google.internal" {
		return true
	}
	ip := net.ParseIP(host)
	if ip == nil {
		// Could be a hostname — resolve it
		addrs, err := net.LookupHost(host)
		if err != nil || len(addrs) == 0 {
			return false // let it fail at HTTP level
		}
		ip = net.ParseIP(addrs[0])
		if ip == nil {
			return false
		}
	}
	return ip.IsLoopback() || ip.IsPrivate() || ip.IsLinkLocalUnicast() || ip.IsLinkLocalMulticast()
}

// LLMProvider holds configuration for a single LLM provider.
type LLMProvider struct {
	Name    string
	BaseURL string
	APIKey  string
	Model   string
}

// LLMMessage is a single chat message.
type LLMMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// LLMRequest is the unified request structure.
type LLMRequest struct {
	Messages    []LLMMessage `json:"messages"`
	MaxTokens   int          `json:"max_tokens,omitempty"`
	Temperature float64      `json:"temperature,omitempty"`
	Stream      bool         `json:"stream,omitempty"`
}

// LLMResponse is the unified response structure.
type LLMResponse struct {
	Content   string `json:"content"`
	Model     string `json:"model"`
	Provider  string `json:"provider"`
	TokensIn  int    `json:"tokens_in"`
	TokensOut int    `json:"tokens_out"`
}

// LLMClient manages multiple LLM providers.
type LLMClient struct {
	providers map[string]*LLMProvider
	http      *http.Client
	logger    *Logger
}

// NewLLMClient creates a new LLMClient with a 30-second timeout.
func NewLLMClient(logger *Logger) *LLMClient {
	return &LLMClient{
		providers: make(map[string]*LLMProvider),
		http:      &http.Client{Timeout: 30 * time.Second},
		logger:    logger,
	}
}

// AddProvider registers a provider by name.
func (c *LLMClient) AddProvider(name, baseURL, apiKey, model string) {
	c.providers[name] = &LLMProvider{
		Name:    name,
		BaseURL: strings.TrimRight(baseURL, "/"),
		APIKey:  apiKey,
		Model:   model,
	}
	c.logger.Debug("LLM provider registered", "provider", name, "model", model)
}

// ListProviders returns the names of all registered providers.
func (c *LLMClient) ListProviders() []string {
	names := make([]string, 0, len(c.providers))
	for name := range c.providers {
		names = append(names, name)
	}
	return names
}

// Complete routes a request to the named provider and returns a unified response.
func (c *LLMClient) Complete(providerName string, req LLMRequest) (*LLMResponse, error) {
	provider, ok := c.providers[providerName]
	if !ok {
		return nil, fmt.Errorf("LLM provider %q not registered", providerName)
	}

	lowerName := strings.ToLower(provider.Name)
	lowerURL := strings.ToLower(provider.BaseURL)

	var providerType string
	switch {
	case strings.Contains(lowerName, "ollama") || strings.Contains(lowerURL, ":11434"):
		providerType = "ollama"
	case strings.Contains(lowerName, "anthropic") || strings.Contains(lowerURL, "anthropic"):
		providerType = "anthropic"
	default:
		providerType = "openai"
	}

	// Allow localhost only for Ollama (local inference)
	if isPrivateIP(provider.BaseURL) && providerType != "ollama" {
		return nil, fmt.Errorf("provider URL points to private network")
	}

	switch providerType {
	case "ollama":
		return c.completeOllama(provider, req)
	case "anthropic":
		return c.completeAnthropic(provider, req)
	default:
		return c.completeOpenAI(provider, req)
	}
}

// ── OpenAI (and OpenAI-compatible) ──────────────────────────────────────────

type openAIRequest struct {
	Model       string       `json:"model"`
	Messages    []LLMMessage `json:"messages"`
	MaxTokens   int          `json:"max_tokens,omitempty"`
	Temperature float64      `json:"temperature,omitempty"`
}

type openAIResponse struct {
	Model   string `json:"model"`
	Choices []struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	} `json:"choices"`
	Usage struct {
		PromptTokens     int `json:"prompt_tokens"`
		CompletionTokens int `json:"completion_tokens"`
	} `json:"usage"`
}

func (c *LLMClient) completeOpenAI(provider *LLMProvider, req LLMRequest) (*LLMResponse, error) {
	payload := openAIRequest{
		Model:       provider.Model,
		Messages:    req.Messages,
		MaxTokens:   req.MaxTokens,
		Temperature: req.Temperature,
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("openai marshal: %w", err)
	}

	url := provider.BaseURL + "/chat/completions"
	httpReq, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("openai new request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")
	if provider.APIKey != "" {
		httpReq.Header.Set("Authorization", "Bearer "+provider.APIKey)
	}
	// OpenRouter requires a referrer header.
	if strings.Contains(strings.ToLower(provider.BaseURL), "openrouter") {
		httpReq.Header.Set("HTTP-Referer", "https://secureyeoman.ai")
	}

	c.logger.Debug("OpenAI request", "provider", provider.Name, "url", url, "model", provider.Model)

	resp, err := c.http.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("openai http: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("openai read body: %w", err)
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("provider request failed (status %d)", resp.StatusCode)
	}

	var parsed openAIResponse
	if err := json.Unmarshal(respBody, &parsed); err != nil {
		return nil, fmt.Errorf("openai parse response: %w", err)
	}
	if len(parsed.Choices) == 0 {
		return nil, fmt.Errorf("openai: no choices in response")
	}

	model := parsed.Model
	if model == "" {
		model = provider.Model
	}

	return &LLMResponse{
		Content:   parsed.Choices[0].Message.Content,
		Model:     model,
		Provider:  provider.Name,
		TokensIn:  parsed.Usage.PromptTokens,
		TokensOut: parsed.Usage.CompletionTokens,
	}, nil
}

// ── Anthropic Messages API ───────────────────────────────────────────────────

type anthropicRequest struct {
	Model       string       `json:"model"`
	Messages    []LLMMessage `json:"messages"`
	MaxTokens   int          `json:"max_tokens"`
	Temperature float64      `json:"temperature,omitempty"`
}

type anthropicResponse struct {
	Model   string `json:"model"`
	Content []struct {
		Text string `json:"text"`
	} `json:"content"`
	Usage struct {
		InputTokens  int `json:"input_tokens"`
		OutputTokens int `json:"output_tokens"`
	} `json:"usage"`
}

func (c *LLMClient) completeAnthropic(provider *LLMProvider, req LLMRequest) (*LLMResponse, error) {
	maxTokens := req.MaxTokens
	if maxTokens == 0 {
		maxTokens = 4096
	}

	payload := anthropicRequest{
		Model:       provider.Model,
		Messages:    req.Messages,
		MaxTokens:   maxTokens,
		Temperature: req.Temperature,
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("anthropic marshal: %w", err)
	}

	// Always use the canonical Anthropic endpoint regardless of the stored base URL
	// (the base URL is used only for provider detection).
	url := "https://api.anthropic.com/v1/messages"
	httpReq, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("anthropic new request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("x-api-key", provider.APIKey)
	httpReq.Header.Set("anthropic-version", "2023-06-01")

	c.logger.Debug("Anthropic request", "provider", provider.Name, "model", provider.Model)

	resp, err := c.http.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("anthropic http: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("anthropic read body: %w", err)
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("provider request failed (status %d)", resp.StatusCode)
	}

	var parsed anthropicResponse
	if err := json.Unmarshal(respBody, &parsed); err != nil {
		return nil, fmt.Errorf("anthropic parse response: %w", err)
	}
	if len(parsed.Content) == 0 {
		return nil, fmt.Errorf("anthropic: no content in response")
	}

	model := parsed.Model
	if model == "" {
		model = provider.Model
	}

	return &LLMResponse{
		Content:   parsed.Content[0].Text,
		Model:     model,
		Provider:  provider.Name,
		TokensIn:  parsed.Usage.InputTokens,
		TokensOut: parsed.Usage.OutputTokens,
	}, nil
}

// ── Ollama ───────────────────────────────────────────────────────────────────

type ollamaRequest struct {
	Model    string       `json:"model"`
	Messages []LLMMessage `json:"messages"`
	Stream   bool         `json:"stream"`
}

type ollamaResponse struct {
	Message struct {
		Content string `json:"content"`
	} `json:"message"`
	Model           string `json:"model"`
	EvalCount       int    `json:"eval_count"`
	PromptEvalCount int    `json:"prompt_eval_count"`
}

func (c *LLMClient) completeOllama(provider *LLMProvider, req LLMRequest) (*LLMResponse, error) {
	payload := ollamaRequest{
		Model:    provider.Model,
		Messages: req.Messages,
		Stream:   false, // always non-streaming for the unified interface
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("ollama marshal: %w", err)
	}

	url := provider.BaseURL + "/api/chat"
	httpReq, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("ollama new request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")

	c.logger.Debug("Ollama request", "provider", provider.Name, "url", url, "model", provider.Model)

	resp, err := c.http.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("ollama http: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("ollama read body: %w", err)
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("provider request failed (status %d)", resp.StatusCode)
	}

	var parsed ollamaResponse
	if err := json.Unmarshal(respBody, &parsed); err != nil {
		return nil, fmt.Errorf("ollama parse response: %w", err)
	}

	model := parsed.Model
	if model == "" {
		model = provider.Model
	}

	return &LLMResponse{
		Content:   parsed.Message.Content,
		Model:     model,
		Provider:  provider.Name,
		TokensIn:  parsed.PromptEvalCount,
		TokensOut: parsed.EvalCount,
	}, nil
}

// ── Auto-configuration ───────────────────────────────────────────────────────

// AutoConfigProviders reads well-known environment variables and registers
// any providers whose credentials are present.
//
// Supported variables:
//
//	OpenAI:      OPENAI_API_KEY, OPENAI_BASE_URL, OPENAI_MODEL
//	Anthropic:   ANTHROPIC_API_KEY, ANTHROPIC_MODEL
//	Ollama:      OLLAMA_URL, OLLAMA_MODEL
//	OpenRouter:  OPENROUTER_API_KEY, OPENROUTER_MODEL
func AutoConfigProviders(logger *Logger) *LLMClient {
	client := NewLLMClient(logger)

	// OpenAI
	if key := os.Getenv("OPENAI_API_KEY"); key != "" {
		baseURL := os.Getenv("OPENAI_BASE_URL")
		if baseURL == "" {
			baseURL = "https://api.openai.com/v1"
		}
		model := os.Getenv("OPENAI_MODEL")
		if model == "" {
			model = "gpt-4o"
		}
		client.AddProvider("openai", baseURL, key, model)
	}

	// Anthropic
	if key := os.Getenv("ANTHROPIC_API_KEY"); key != "" {
		model := os.Getenv("ANTHROPIC_MODEL")
		if model == "" {
			model = "claude-sonnet-4-20250514"
		}
		client.AddProvider("anthropic", "https://api.anthropic.com/v1", key, model)
	}

	// Ollama (no API key required)
	ollamaURL := os.Getenv("OLLAMA_URL")
	if ollamaURL == "" {
		ollamaURL = "http://localhost:11434"
	}
	ollamaModel := os.Getenv("OLLAMA_MODEL")
	if ollamaModel == "" {
		ollamaModel = "llama3.2"
	}
	// Only register Ollama if the URL was explicitly set or we can assume local presence.
	// Register unconditionally so callers can attempt it; failures surface at call time.
	client.AddProvider("ollama", ollamaURL, "", ollamaModel)

	// OpenRouter
	if key := os.Getenv("OPENROUTER_API_KEY"); key != "" {
		model := os.Getenv("OPENROUTER_MODEL")
		if model == "" {
			model = "anthropic/claude-sonnet-4-20250514"
		}
		client.AddProvider("openrouter", "https://openrouter.ai/api/v1", key, model)
	}

	logger.Info("LLM auto-config complete", "providers", strings.Join(client.ListProviders(), ","))
	return client
}

File: crates/sy-edge/src/llm.rs
//! LLM client — multi-provider completion requests (OpenAI, Anthropic, Ollama, OpenRouter).

use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct ProviderInfo {
    pub name: String,
    pub model: String,
    pub configured: bool,
}

#[derive(Clone)]
pub struct LlmClient {
    providers: Vec<ProviderConfig>,
    http: reqwest::Client,
}

#[derive(Clone)]
struct ProviderConfig {
    name: String,
    api_key: String,
    base_url: String,
    model: String,
}

impl LlmClient {
    pub fn from_env() -> Self {
        let mut providers = Vec::new();

        // OpenAI
        if let Ok(key) = std::env::var("OPENAI_API_KEY") {
            providers.push(ProviderConfig {
                name: "openai".into(),
                api_key: key,
                base_url: std::env::var("OPENAI_BASE_URL")
                    .unwrap_or_else(|_| "https://api.openai.com/v1".into()),
                model: std::env::var("OPENAI_MODEL").unwrap_or_else(|_| "gpt-4o-mini".into()),
            });
        }

        // Anthropic
        if let Ok(key) = std::env::var("ANTHROPIC_API_KEY") {
            providers.push(ProviderConfig {
                name: "anthropic".into(),
                api_key: key,
                base_url: "https://api.anthropic.com/v1".into(),
                model: std::env::var("ANTHROPIC_MODEL")
                    .unwrap_or_else(|_| "claude-sonnet-4-20250514".into()),
            });
        }

        // Ollama
        if let Ok(url) = std::env::var("OLLAMA_URL") {
            providers.push(ProviderConfig {
                name: "ollama".into(),
                api_key: String::new(),
                base_url: url,
                model: std::env::var("OLLAMA_MODEL").unwrap_or_else(|_| "llama3.2".into()),
            });
        }

        // OpenRouter
        if let Ok(key) = std::env::var("OPENROUTER_API_KEY") {
            providers.push(ProviderConfig {
                name: "openrouter".into(),
                api_key: key,
                base_url: "https://openrouter.ai/api/v1".into(),
                model: std::env::var("OPENROUTER_MODEL")
                    .unwrap_or_else(|_| "anthropic/claude-sonnet-4-20250514".into()),
            });
        }

        Self {
            providers,
            http: reqwest::Client::builder()
                .timeout(std::time::Duration::from_secs(60))
                .build()
                .unwrap_or_default(),
        }
    }

    pub async fn complete(
        &self,
        prompt: &str,
        provider_name: Option<&str>,
        model_override: Option<&str>,
        max_tokens: u32,
    ) -> Result<String, String> {
        let provider = if let Some(name) = provider_name {
            self.providers
                .iter()
                .find(|p| p.name == name)
                .ok_or_else(|| format!("Provider not found: {name}"))?
        } else {
            self.providers
                .first()
                .ok_or("No LLM providers configured")?
        };

        let model = model_override.unwrap_or(&provider.model);

        match provider.name.as_str() {
            "anthropic" => self.complete_anthropic(provider, prompt, model, max_tokens).await,
            "ollama" => self.complete_ollama(provider, prompt, model).await,
            _ => self.complete_openai_compat(provider, prompt, model, max_tokens).await,
        }
    }

    async fn complete_openai_compat(
        &self,
        provider: &ProviderConfig,
        prompt: &str,
        model: &str,
        max_tokens: u32,
    ) -> Result<String, String> {
        let body = serde_json::json!({
            "model": model,
            "messages": [{ "role": "user", "content": prompt }],
            "max_tokens": max_tokens,
        });

        let resp = self
            .http
            .post(format!("{}/chat/completions", provider.base_url))
            .bearer_auth(&provider.api_key)
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("Request failed: {e}"))?;

        if !resp.status().is_success() {
            let text = resp.text().await.unwrap_or_default();
            return Err(format!("API error: {text}"));
        }

        let data: serde_json::Value = resp.json().await.map_err(|e| format!("Parse error: {e}"))?;
        Ok(data["choices"][0]["message"]["content"]
            .as_str()
            .unwrap_or("")
            .to_string())
    }

    async fn complete_anthropic(
        &self,
        provider: &ProviderConfig,
        prompt: &str,
        model: &str,
        max_tokens: u32,
    ) -> Result<String, String> {
        let body = serde_json::json!({
            "model": model,
            "messages": [{ "role": "user", "content": prompt }],
            "max_tokens": max_tokens,
        });

        let resp = self
            .http
            .post(format!("{}/messages", provider.base_url))
            .header("x-api-key", &provider.api_key)
            .header("anthropic-version", "2023-06-01")
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("Request failed: {e}"))?;

        if !resp.status().is_success() {
            let text = resp.text().await.unwrap_or_default();
            return Err(format!("API error: {text}"));
        }

        let data: serde_json::Value = resp.json().await.map_err(|e| format!("Parse error: {e}"))?;
        Ok(data["content"][0]["text"]
            .as_str()
            .unwrap_or("")
            .to_string())
    }

    async fn complete_ollama(
        &self,
        provider: &ProviderConfig,
        prompt: &str,
        model: &str,
    ) -> Result<String, String> {
        let body = serde_json::json!({
            "model": model,
            "prompt": prompt,
            "stream": false,
        });

        let resp = self
            .http
            .post(format!("{}/api/generate", provider.base_url))
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("Request failed: {e}"))?;

        if !resp.status().is_success() {
            let text = resp.text().await.unwrap_or_default();
            return Err(format!("API error: {text}"));
        }

        let data: serde_json::Value = resp.json().await.map_err(|e| format!("Parse error: {e}"))?;
        Ok(data["response"].as_str().unwrap_or("").to_string())
    }

    pub fn provider_count(&self) -> usize {
        self.providers.len()
    }

    pub fn list_providers(&self) -> Vec<ProviderInfo> {
        self.providers
            .iter()
            .map(|p| ProviderInfo {
                name: p.name.clone(),
                model: p.model.clone(),
                configured: true,
            })
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn list_providers_returns_vec() {
        // Just verify from_env doesn't panic and returns a valid vec
        let client = LlmClient::from_env();
        let providers = client.list_providers();
        for p in &providers {
            assert!(!p.name.is_empty());
            assert!(p.configured);
        }
    }

    #[test]
    fn from_env_with_openai() {
        unsafe { std::env::set_var("OPENAI_API_KEY", "sk-test-key") };
        let client = LlmClient::from_env();
        assert!(client.provider_count() >= 1);
        let providers = client.list_providers();
        assert!(providers.iter().any(|p| p.name == "openai"));
        unsafe { std::env::remove_var("OPENAI_API_KEY") };
    }

    #[test]
    fn from_env_with_ollama() {
        unsafe { std::env::set_var("OLLAMA_URL", "http://localhost:11434") };
        let client = LlmClient::from_env();
        let providers = client.list_providers();
        assert!(providers.iter().any(|p| p.name == "ollama"));
        unsafe { std::env::remove_var("OLLAMA_URL") };
    }

    #[test]
    fn provider_info_serialization() {
        let info = ProviderInfo {
            name: "openai".into(),
            model: "gpt-4o-mini".into(),
            configured: true,
        };
        let json = serde_json::to_string(&info).unwrap();
        assert!(json.contains("\"name\":\"openai\""));
    }

    #[test]
    fn custom_model_via_env() {
        unsafe { std::env::set_var("OPENAI_API_KEY", "sk-test") };
        unsafe { std::env::set_var("OPENAI_MODEL", "gpt-4o") };
        let client = LlmClient::from_env();
        let providers = client.list_providers();
        let openai = providers.iter().find(|p| p.name == "openai").unwrap();
        assert_eq!(openai.model, "gpt-4o");
        unsafe { std::env::remove_var("OPENAI_API_KEY") };
        unsafe { std::env::remove_var("OPENAI_MODEL") };
    }
}
